### PR TITLE
Generate correct types according to https://arethetypeswrong.github.io

### PR DESCRIPTION
**Pull request by mrm007**
Same story as https://github.com/seek-oss/crackle/pull/156.

Current results: https://arethetypeswrong.github.io/?p=braid-design-system@32.14.1
Snapshot results: https://arethetypeswrong.github.io/?p=braid-design-system@0.0.0-are-the-types-wrong-20231206063811

The [**Incorrect default export**](https://github.com/arethetypeswrong/arethetypeswrong.github.io/blob/main/docs/problems/FalseExportDefault.md) error is a little trickier to fix, and might require a breaking change.

---

I tested in a `sku init` repo to make sure types are resolved correctly in a real project.

<details>
<summary>moduleResolution: node (legacy)</summary>

```
yarn run v1.22.19
$ /___/node_modules/.bin/tsc --moduleResolution node --traceResolution src/App/NextSteps.tsx
======== Resolving module 'braid-design-system' from '/___/src/App/NextSteps.tsx'. ========
Explicitly specified module resolution kind: 'Node10'.
Loading module 'braid-design-system' from 'node_modules' folder, target file types: TypeScript, Declaration.
Searching all ancestor node_modules directories for preferred extensions: TypeScript, Declaration.
Directory '/___/src/App/node_modules' does not exist, skipping all lookups in it.
Directory '/___/src/node_modules' does not exist, skipping all lookups in it.
Found 'package.json' at '/___/node_modules/braid-design-system/package.json'.
File '/___/node_modules/braid-design-system.ts' does not exist.
File '/___/node_modules/braid-design-system.tsx' does not exist.
File '/___/node_modules/braid-design-system.d.ts' does not exist.
'package.json' does not have a 'typesVersions' field.
'package.json' does not have a 'typings' field.
'package.json' has 'types' field './dist/index.d.ts' that references '/___/node_modules/braid-design-system/dist/index.d.ts'.
File '/___/node_modules/braid-design-system/dist/index.d.ts' exists - use it as a name resolution result.
Resolving real path for '/___/node_modules/braid-design-system/dist/index.d.ts', result '/___/node_modules/braid-design-system/dist/index.d.ts'.
======== Module name 'braid-design-system' was successfully resolved to '/___/node_modules/braid-design-system/dist/index.d.ts' with Package ID 'braid-design-system/dist/index.d.ts@0.0.0-are-the-types-wrong-20231206063811'. ========
```

</details>

<details>
<summary>moduleResolution: bundler (default from sku)</summary>

```
yarn run v1.22.19
$ /___/node_modules/.bin/tsc --moduleResolution bundler --traceResolution src/App/NextSteps.tsx
======== Resolving module 'braid-design-system' from '/___/src/App/NextSteps.tsx'. ========
Explicitly specified module resolution kind: 'Bundler'.
Resolving in CJS mode with conditions 'import', 'types'.
File '/___/src/App/package.json' does not exist.
File '/___/src/package.json' does not exist.
Found 'package.json' at '/___/package.json'.
Loading module 'braid-design-system' from 'node_modules' folder, target file types: TypeScript, JavaScript, Declaration, JSON.
Searching all ancestor node_modules directories for preferred extensions: TypeScript, Declaration.
Directory '/___/src/App/node_modules' does not exist, skipping all lookups in it.
Directory '/___/src/node_modules' does not exist, skipping all lookups in it.
Found 'package.json' at '/___/node_modules/braid-design-system/package.json'.
Entering conditional exports.
Matched 'exports' condition 'types'.
Entering conditional exports.
Matched 'exports' condition 'import'.
Using 'exports' subpath '.' with target './dist/index.d.mts'.
File '/___/node_modules/braid-design-system/dist/index.d.mts' exists - use it as a name resolution result.
Resolved under condition 'import'.
Exiting conditional exports.
Resolved under condition 'types'.
Exiting conditional exports.
Resolving real path for '/___/node_modules/braid-design-system/dist/index.d.mts', result '/___/node_modules/braid-design-system/dist/index.d.mts'.
======== Module name 'braid-design-system' was successfully resolved to '/___/node_modules/braid-design-system/dist/index.d.mts' with Package ID 'braid-design-system/dist/index.d.mts@0.0.0-are-the-types-wrong-20231206063811'. ========

```

</details>

<details>
<summary>moduleResolution: node16 (package.json implicit type: commonjs)</summary>

```
yarn run v1.22.19
$ /___/node_modules/.bin/tsc --moduleResolution node16 --traceResolution src/App/NextSteps.tsx
File '/___/src/App/package.json' does not exist.
File '/___/src/package.json' does not exist.
Found 'package.json' at '/___/package.json'.
======== Resolving module 'braid-design-system' from '/___/src/App/NextSteps.tsx'. ========
Explicitly specified module resolution kind: 'Node16'.
Resolving in CJS mode with conditions 'require', 'types', 'node'.
File '/___/src/App/package.json' does not exist according to earlier cached lookups.
File '/___/src/package.json' does not exist according to earlier cached lookups.
File '/___/package.json' exists according to earlier cached lookups.
Loading module 'braid-design-system' from 'node_modules' folder, target file types: TypeScript, JavaScript, Declaration.
Searching all ancestor node_modules directories for preferred extensions: TypeScript, Declaration.
Directory '/___/src/App/node_modules' does not exist, skipping all lookups in it.
Directory '/___/src/node_modules' does not exist, skipping all lookups in it.
Found 'package.json' at '/___/node_modules/braid-design-system/package.json'.
Entering conditional exports.
Matched 'exports' condition 'types'.
Entering conditional exports.
Saw non-matching condition 'import'.
Matched 'exports' condition 'require'.
Using 'exports' subpath '.' with target './dist/index.d.ts'.
File '/___/node_modules/braid-design-system/dist/index.d.ts' exists - use it as a name resolution result.
Resolved under condition 'require'.
Exiting conditional exports.
Resolved under condition 'types'.
Exiting conditional exports.
Resolving real path for '/___/node_modules/braid-design-system/dist/index.d.ts', result '/___/node_modules/braid-design-system/dist/index.d.ts'.
======== Module name 'braid-design-system' was successfully resolved to '/___/node_modules/braid-design-system/dist/index.d.ts' with Package ID 'braid-design-system/dist/index.d.ts@0.0.0-are-the-types-wrong-20231206063811'. ========
```

</details>

<details>
<summary>moduleResolution node16 (package.json type: module)</summary>

```
yarn run v1.22.19
$ /___/node_modules/.bin/tsc --moduleResolution node16 --traceResolution src/App/NextSteps.tsx
File '/___/src/App/package.json' does not exist.
File '/___/src/package.json' does not exist.
Found 'package.json' at '/___/package.json'.
======== Resolving module 'braid-design-system' from '/___/src/App/NextSteps.tsx'. ========
Explicitly specified module resolution kind: 'Node16'.
Resolving in ESM mode with conditions 'import', 'types', 'node'.
File '/___/src/App/package.json' does not exist according to earlier cached lookups.
File '/___/src/package.json' does not exist according to earlier cached lookups.
File '/___/package.json' exists according to earlier cached lookups.
Loading module 'braid-design-system' from 'node_modules' folder, target file types: TypeScript, JavaScript, Declaration.
Searching all ancestor node_modules directories for preferred extensions: TypeScript, Declaration.
Directory '/___/src/App/node_modules' does not exist, skipping all lookups in it.
Directory '/___/src/node_modules' does not exist, skipping all lookups in it.
Found 'package.json' at '/___/node_modules/braid-design-system/package.json'.
Entering conditional exports.
Matched 'exports' condition 'types'.
Entering conditional exports.
Matched 'exports' condition 'import'.
Using 'exports' subpath '.' with target './dist/index.d.mts'.
File '/___/node_modules/braid-design-system/dist/index.d.mts' exists - use it as a name resolution result.
Resolved under condition 'import'.
Exiting conditional exports.
Resolved under condition 'types'.
Exiting conditional exports.
Resolving real path for '/___/node_modules/braid-design-system/dist/index.d.mts', result '/___/node_modules/braid-design-system/dist/index.d.mts'.
======== Module name 'braid-design-system' was successfully resolved to '/___/node_modules/braid-design-system/dist/index.d.mts' with Package ID 'braid-design-system/dist/index.d.mts@0.0.0-are-the-types-wrong-20231206063811'. ========
```

</details>

**TL;DR**: It works.

### CODE REVIEW

File: packages/generate-component-docs/src/__snapshots__/contract.test.ts.snap [hunk 1]
```diff
@@ -8214,7 +8214,7 @@ exports[`filterSuggestions 1`] = `
   props: {
     at: (index: number) => FilterableSuggestion<Value> | FilterableGroupedSuggestions<Value> | undefined
     concat: { (...items: ConcatArray<FilterableSuggestion<Value> | FilterableGroupedSuggestions<Value>>[]): (FilterableSuggestion<...> | FilterableGroupedSuggestions<...>)[]; (...items: (FilterableSuggestion<...> | ... 1 more ... | ConcatArray<...>)[]): (FilterableSuggestion<...> | FilterableGroupedSuggestions<...>)[]; }
-    copyWithin: (target: number, start?: number | undefined, end?: number | undefined) => FilterableSuggestions<Value>
+    copyWithin: (target: number, start: number, end?: number | undefined) => FilterableSuggestions<Value>
```
Comment (mrm007) [inline packages/generate-component-docs/src/__snapshots__/contract.test.ts.snap:8217]: `filterSuggestions`'s argument is an array. TypeScript changed, so the type definition for `Array` must have changed as well.

This is a problem with how we extract type information from component exports.

File: .changeset/twenty-badgers-hide.md [hunk 2]
```diff
@@ -0,0 +1,6 @@
+---
+'braid-design-system': patch
+'@braid-design-system/docs-ui': patch
+---
+
+Now providing types correctly according to https://arethetypeswrong.github.io
```
Comment (askoufis) [inline .changeset/twenty-badgers-hide.md:6]: More rewording 😅.
```suggestion
Provide correct types according to https://arethetypeswrong.github.io
```

**Review by askoufis — APPROVED** [review]
